### ISSUE TITLE
Installing randoop 4.2.1 I get Error: Could not find or load main class randoop.main.Main

### ISSUE THREAD

**Issue by rickersilva** [randoop/randoop#644]
I´m installing Randoop 4.2.1 but cannot run it.  I have two zip files for randoop, the one here 
https://github.com/randoop/randoop/releases/download/v4.2.1/randoop-4.2.1.zip

and the one here
https://github.com/randoop/randoop/archive/v4.2.1.zip

which was given to me in this post (https://github.com/randoop/randoop/issues/332) where I started asking about this issue.

The first one I unzipped in c:\randoop-4.2.1 and the second one in c:\randoop-4.2.1-full

Content of each of the installations are very different:
 Directory: C:\randoop-4.2.1

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        1/10/2020  10:03 AM        6819062 covered-class-4.2.1.jar
-a----        1/10/2020  10:03 AM        5957734 randoop-all-4.2.1.jar
-a----        1/10/2020  10:03 AM            151 README.txt
-a----        1/10/2020  10:03 AM        2078938 replacecall-4.2.1.jar

  Directory: C:\randoop-4.2.1-full

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        1/13/2020   3:02 PM                .circleci
d-----        1/13/2020   3:02 PM                .github
d-----        1/13/2020   3:02 PM                agent
d-----        1/13/2020   3:02 PM                gradle
d-----        1/13/2020   3:02 PM                scripts
d-----        1/13/2020   3:03 PM                src
-a----        1/13/2020   3:03 PM            415 .gitignore
-a----        1/13/2020   3:03 PM           4842 .travis-build.sh
-a----        1/13/2020   3:03 PM           3645 .travis.yml
-a----        1/13/2020   3:03 PM           1355 azure-pipelines.yml
-a----        1/13/2020   3:03 PM          36064 build.gradle
-a----        1/13/2020   3:03 PM            441 checkstyle.xml
-a----        1/13/2020   3:03 PM           5296 gradlew
-a----        1/13/2020   3:03 PM           2260 gradlew.bat
-a----        1/13/2020   3:03 PM           2194 license.txt
-a----        1/13/2020   3:03 PM           1676 README.md
-a----        1/13/2020   3:03 PM            807 settings.gradle
 
Once installed I run the command in the manual: java -Xmx3000m -classpath myclasspath:$(RANDOOP_JAR) randoop.main.Main gentests --testclass=java.util.TreeSet --output-limit=100

Which in my case is like this:
**java -Xmx3000m -classpath C:\[HIDDEN PATH TO CLASS FILE]\ClassFile.class:$([RANDOOP ROOT]\randoop-all-4.2.1.jar) randoop.main.Main gentests --testclass=java.util.TreeSet --output-limit=100**
the result placed on the randoop-4.2.1 folder is: Error: Could not find or load main class randoop.main.Main
the result placed on the randoop-4.2.1-full is: Error: Could not find or load main class randoop.main.Main

Where the **[HIDDEN PATH TO CLASS FILE]** points to the .class file of the class I need the tests to be generated and is located on a source folder within my windows user folder.
Where **[RANDOOP ROOT]** is the folder where I unzipped (installed) randoop for both zip files I mentioned previously.
The rest is exactly as stated in Randoop Manual

As it did not run, I tried one of the help commands also in the manual. This one: 
java -classpath $(RANDOOP_JAR) randoop.main.Main help
and the result was:
on randoop-4.2.1\ folder: 
Error: Could not find or load main class help

on randoop-4.2.1-full\ folder:
C:\randoop-4.2.1-full\randoop-all-4.2.1.jar : The term 'C:\randoop-4.2.1-full\randoop-all-4.2.1.jar' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:19
+ java -classpath $(C:\randoop-4.2.1-full\randoop-all-4.2.1.jar) randoo ...
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\randoop-4.2....p-all-4.2.1.jar:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

Error: Could not find or load main class help

Which is expected since that folder has no jar files. 

In the troubleshoot section of the manual there are some indications for this errors. I followed them at my best but still can't get it to run.

I also double checked my environment variables and are as listed by the reg command as follows

PS C:\randoop-4.2.1-full> reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment"

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment
    **JAVA_HOME    REG_SZ    C:\Program Files\Java\jre1.8.0_231**
    **Path    REG_EXPAND_SZ    %JAVA_HOME%\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;%SystemRoot%\system32;%SystemRoot%;%SystemRoot%\System32\Wbem;%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\;%SYSTEMROOT%\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;**
    PATHEXT    REG_SZ    .COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
    USERNAME    REG_SZ    SYSTEM
    windir    REG_EXPAND_SZ    %SystemRoot%

PS C:\randoop-4.2.1-full> reg query HKCU\Environment
>>

HKEY_CURRENT_USER\Environment
    OneDrive    REG_EXPAND_SZ    C:\Users\s1878883\OneDrive
    **Path    REG_EXPAND_SZ    %JAVA_HOME%\bin;%USERPROFILE%\AppData\Local\Microsoft\WindowsApps;**
    TEMP    REG_EXPAND_SZ    %USERPROFILE%\AppData\Local\Temp
    TMP    REG_EXPAND_SZ    %USERPROFILE%\AppData\Local\Temp

What I expected to happen instead?
I expected to unpack the zip file as it is linked in the manual, run the command described in the manual in console with the path to the class file I wnat the test to be written to, and receive confirmation of the process.

If you can guide me into this. I have not much experience with java but I have some guys beside me that will help me with that in case I need.

**Comment by CrazyCoder** [comment]
The problem is with the syntax. You don't use parentheses in the command line classpath on Windows.

Try this instead:

```
java -Xmx3000m -classpath d:\dev\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main gentests --testclass=java.util.TreeSet --output-limit=100
```

Note that you need JDK, not JRE, I've explained it here: https://stackoverflow.com/a/59690659/104891 .

JRE doesn't have tools.jar required for randoop to work.

If you need to add your own classes to the classpath, use the semicolon to separate the paths, refer to the `java` command line syntax documentation: https://docs.oracle.com/javase/7/docs/technotes/tools/windows/classpath.html .

Note that classpath points to the directories or jars, not individual `.class` files.

**Comment by rickersilva** [comment]
Thank you for your answer. I don't understand this

> Note that classpath points to the directories or jars, not individual .class files.

What it means? My projects are quite complex and not all classes need unit tests. 

How can I indicate randoop which classes I want to create unit tests for?

Is this the purpose of the text file some people refer to?

**Comment by CrazyCoder** [comment]
> How can I indicate randoop which classes I want to create unit tests for?

I guess `--testclass` option does that?

I'm sure there are other way to limit the classes, refer to the docs.

**Comment by rickersilva** [comment]
Here https://stackoverflow.com/questions/59690071/java-lang-illegalstateexception-trying-to-generate-unit-tests-with-randoop/59690659#59690659

you say, and I quote "Double check the actual PATH. It may point to JRE instead of the JDK and the System PATH has priority over the User PATH."
I´m pointing to JRE.

but in the post above you said: "Note that you need JDK, not JRE,"

I´m confused but i did change it again back to JDK and it still does not run.

**PS C:\randoop-4.2.1> java -Xmx3000m -classpath C:\Users\s1878883\Source\gcms-baas-card\gateway\build\classes\java\main\com\bns\baas\card\gateway\account\CreditCardFindAll.class;C:\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main gentests --testclass=java.util.TreeSet --output-limit=100**
Usage: java [-options] class [args...]
           (to execute a class)
   or  java [-options] -jar jarfile [args...]
           (to execute a jar file)
where options include:
    -d32          use a 32-bit data model if available
    -d64          use a 64-bit data model if available
    -server       to select the "server" VM
                  The default VM is server.

    -cp <class search path of directories and zip/jar files>
    -classpath <class search path of directories and zip/jar files>
                  A ; separated list of directories, JAR archives,
                  and ZIP archives to search for class files.
    -D<name>=<value>
                  set a system property
    -verbose:[class|gc|jni]
                  enable verbose output
    -version      print product version and exit
    -version:<value>
                  Warning: this feature is deprecated and will be removed
                  in a future release.
                  require the specified version to run
    -showversion  print product version and continue
    -jre-restrict-search | -no-jre-restrict-search
                  Warning: this feature is deprecated and will be removed
                  in a future release.
                  include/exclude user private JREs in the version search
    -? -help      print this help message
    -X            print help on non-standard options
    -ea[:<packagename>...|:<classname>]
    -enableassertions[:<packagename>...|:<classname>]
                  enable assertions with specified granularity
    -da[:<packagename>...|:<classname>]
    -disableassertions[:<packagename>...|:<classname>]
                  disable assertions with specified granularity
    -esa | -enablesystemassertions
                  enable system assertions
    -dsa | -disablesystemassertions
                  disable system assertions
    -agentlib:<libname>[=<options>]
                  load native agent library <libname>, e.g. -agentlib:hprof
                  see also, -agentlib:jdwp=help and -agentlib:hprof=help
    -agentpath:<pathname>[=<options>]
                  load native agent library by full pathname
    -javaagent:<jarpath>[=<options>]
                  load Java programming language agent, see java.lang.instrument
    -splash:<imagepath>
                  show splash screen with specified image
See http://www.oracle.com/technetwork/java/javase/documentation/index.html for more details.
**PS C:\randoop-4.2.1> java -classpath C:\randoop-4.2.1-full\randoop-all-4.2.1.jar randoop.main.Main help**
Error: Could not find or load main class randoop.main.Main

This is the environment variables I used

PS C:\randoop-4.2.1> reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment"

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment
    JAVA_HOME    REG_SZ    C:\Program Files\Java\jdk1.8.0_231
    Path    REG_EXPAND_SZ    C:\Program Files\Java\jdk1.8.0_231\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;%SystemRoot%\system32;%SystemRoot%;%SystemRoot%\System32\Wbem;%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\;%SYSTEMROOT%\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;
    USERNAME    REG_SZ    SYSTEM
    windir    REG_EXPAND_SZ    %SystemRoot%

PS C:\randoop-4.2.1> reg query HKCU\Environment
>>

HKEY_CURRENT_USER\Environment
    JAVA_HOME    REG_SZ    C:\Program Files\Java\jdk1.8.0_231
    Path    REG_EXPAND_SZ    C:\Program Files\Java\jdk1.8.0_231\bin;%USERPROFILE%\AppData\Local\Microsoft\WindowsApps;

**Comment by CrazyCoder** [comment]
>  It may point to JRE instead of the JDK 

Sorry if it was confusing, but that was a suggestion it's the reason of the problem.  If your path points to JRE instead of the JDK, you will have the problem. `may point` means that in your current settings it may point to JRE instead of JDK which is causing the issue, so the solution is to make it point to JDK instead of the JRE.

> Error: Could not find or load main class randoop.main.Main

`C:\randoop-4.2.1-full` contains sources, the jars are in `C:\randoop-4.2.1` according to what you have provided above.

Run this instead:

```
java -classpath C:\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main help
```

`C:\randoop-4.2.1-full\randoop-all-4.2.1.jar` doesn't exist on your disk, it's `C:\randoop-4.2.1\randoop-all-4.2.1.jar` instead!

**Comment by rickersilva** [comment]
This feels really frustrating but It hink I´m getting closer.

running this, seems to be working
PS C:\randoop-4.2.1> java -classpath C:\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main help
Randoop for Java version 4.2.1.

Randoop is a command-line tool that creates unit tests for Java.
It accepts one of the commands listed below. For the user manual,
please visit https://randoop.github.io/randoop/manual/index.html

Type `help' followed by a command name to see documentation.


Commands:

gentests -- Generates unit tests for a set of classes.

help -- Displays a help message for a given command.

minimize -- Minimize a failing JUnit test suite.

Howwever, when I ran this, java -Xmx3000m -classpath C:\Users\s1878883\Source\gcms-baas-card\gateway\build\classes\java\main\com\bns\baas\card\gateway\account\;C:\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main gentests --testclass=CreditCardFindAll --output-limit=100

I get this:

Usage: java [-options] class [args...]
           (to execute a class)
   or  java [-options] -jar jarfile [args...]
           (to execute a jar file)
where options include:
    -d32          use a 32-bit data model if available
    -d64          use a 64-bit data model if available
    -server       to select the "server" VM
                  The default VM is server.

    -cp <class search path of directories and zip/jar files>
    -classpath <class search path of directories and zip/jar files>
                  A ; separated list of directories, JAR archives,
                  and ZIP archives to search for class files.
    -D<name>=<value>
                  set a system property
    -verbose:[class|gc|jni]
                  enable verbose output
    -version      print product version and exit
    -version:<value>
                  Warning: this feature is deprecated and will be removed
                  in a future release.
                  require the specified version to run
    -showversion  print product version and continue
    -jre-restrict-search | -no-jre-restrict-search
                  Warning: this feature is deprecated and will be removed
                  in a future release.
                  include/exclude user private JREs in the version search
    -? -help      print this help message
    -X            print help on non-standard options
    -ea[:<packagename>...|:<classname>]
    -enableassertions[:<packagename>...|:<classname>]
                  enable assertions with specified granularity
    -da[:<packagename>...|:<classname>]
    -disableassertions[:<packagename>...|:<classname>]
                  disable assertions with specified granularity
    -esa | -enablesystemassertions
                  enable system assertions
    -dsa | -disablesystemassertions
                  disable system assertions
    -agentlib:<libname>[=<options>]
                  load native agent library <libname>, e.g. -agentlib:hprof
                  see also, -agentlib:jdwp=help and -agentlib:hprof=help
    -agentpath:<pathname>[=<options>]
                  load native agent library by full pathname
    -javaagent:<jarpath>[=<options>]
                  load Java programming language agent, see java.lang.instrument
    -splash:<imagepath>
                  show splash screen with specified image
See http://www.oracle.com/technetwork/java/javase/documentation/index.html for more details.

To me it seems it is using the correct java. However I´m not 100% sure. The problem is syntax, and I tryed also this.

I created a myclasses.txt in the randoop folder with the names of the classes to test and run this, according to manual:
java -classpath C:\randoop-4.2.1\randoop-all-4.2.1.jar randoop.main.Main gentests --classlist=myclasses.txt --time-limit=60
RESULT
Throwable thrown while handling command: java.lang.IllegalStateException: Cannot find the Java compiler. Check that classpath includes tools.jar
java.lang.IllegalStateException: Cannot find the Java compiler. Check that classpath includes tools.jar
        at randoop.compile.SequenceCompiler.<init>(SequenceCompiler.java:64)
        at randoop.compile.SequenceCompiler.<init>(SequenceCompiler.java:48)
        at randoop.condition.SpecificationCollection.<init>(SpecificationCollection.java:82)
        at randoop.condition.SpecificationCollection.create(SpecificationCollection.java:102)
        at randoop.main.GenTests.handle(GenTests.java:279)
        at randoop.main.Main.nonStaticMain(Main.java:66)
        at randoop.main.Main.main(Main.java:30)

Randoop failed.
No sequences generated.

I understand there is no way randoop cpuld know where the classes are because in the myclasses file there is no location of the classes, nor it is on the sommand to run. So I think I need more information on how to do that.

**Comment by CrazyCoder** [comment]
I guess you need to run something like this:

```
java -Xmx3000m -classpath "C:\Users\s1878883\Source\gcms-baas-card\gateway\build\classes\java\main;C:\randoop-4.2.1\randoop-all-4.2.1.jar" randoop.main.Main gentests --testclass=com.bns.baas.card.gateway.account.CreditCardFindAll --output-limit=100
```

`-classpath` specifies the root package for your classes (not the directory of the .class file) and `--testclass` specifies the fully qualified class name (which includes the package).

Entries inside `-classpath` argument must be included in the double quotes, or the semicolon gets parsed by the shell.

**Comment by mernst** [comment]
@rickersilva What is your current state?  Did the latest suggestion from @CrazyCoder work, or are you still stuck?  Also, what version of Java are you using (8 or 11)?

[A week ago](https://github.com/randoop/randoop/commit/a0eb130a11770d4f3edaab3ef4e27e06a0ce89ff) I improved the "Cannot find the Java compiler." error message, so that now it prints the classpath.  If you are comfortable doing so, you could build Randoop and use the latest version to get a more detailed diagnostic.  But please let us know your status regardless.

**Comment by mernst** [comment]
@rickersilva Ping.

**Comment by rickersilva** [comment]
Sorry I wasn´t online the past two weeks, I´m back working on it. I´ll let you knoe the progress later. Please, can you remove the path in the last answerr you gave me. I mistakenly put information that I don't want to be available.

**Comment by mernst** [comment]
Was the latter part of your answer directed to @CrazyCoder?

**Comment by rickersilva** [comment]
for both.

**Comment by mernst** [comment]
@rickersilva I don't see any paths in my answers in this thread.  Can you point me at the comment of mine that contains a path?

**Comment by rickersilva** [comment]
Oh I see. that part is for @CrazyCoder .

**Comment by rickersilva** [comment]
I´m running it in a developer machine, not mine, since I was facing other problems, now I´n dealing again with the java compiler not found and figuring this out. i find this  a bit weird since this is an actual work station and the project is being developed in this machine and it just runs. 

Is there any way to set the jdk I want to use on the rnadoop gentests command?

**Comment by rickersilva** [comment]
Well, past this java compiler thing. I'm stuck again here.

I run the command

`java -classpath "D:\repositorio\[PATH TO CLASSES MAIN FOLDER];C:\randoop-4.2.1\randoop-all-4.2.1.jar" randoop.main.Main gentests --testclass=[FULLY QUALIFIED PACKAGE].CreditCardFindAll`

and get the following error

`Error: Unable to load class "[FULLY QUALIFIED PACKAGE].CreditCardFindAll" due to exception: java.lang.ClassNotFoundException: [CUSTOM EXCEPTIONS HIERARCHY].exceptions.NotFoundException`

The project is running successfully in the machine. what can cause that kind of exception? Powershell runs as administrator.

**Comment by mernst** [comment]
@rickersilva I would ilke to help, but without a stack trace I don't know what is trying to load `CreditCardFindAll`.
Also, without knowing the classpath I cannot diagnose errors in setting the classpath, but that is almost certainly the problem.

@CrazyCoder pointed out 17 days ago:
> Note that classpath points to the directories or jars, not individual .class files.
That is, elements of the classpath are either directories that contain .class files, or are .jar files that contain .class files.

**Comment by rickersilva** [comment]
Yes, that was it. all jar and classes references were added to the classpath and it run. I have my first set of unit tests. for the record and others that come and check, this is the command I used.

java -classpath "[PATH TO CLASS TO TEST FOLDER];[PATHS TO REFERENCED CLASSES FOLDERS];[PATHS TO REFERENCED JAR FILES]" randoop.main.Main gentests --testclass=[FULLY.QUALIFIED.CLASS_NAME_TO_TEST]

where CLASS TO TEST refers the class I want to create unit tests for. All paths to classes folders or jar are separated by semicolon.
